thetis/rungekutta.py

```diff
@@ -558,6 +558,132 @@ def solve_stage(self, i_stage, t, update_forcings=None):
+            u_nl = u
+        bnd = bnd_conditions
+        fields = self.fields
+
+        # construct variational problems for each stage
+        self.F = []
+        for i in range(self.n_stages):
+            mass = self.equation.mass_term(u) - self.equation.mass_term(u_old)
+            rhs = self.dt_const*self.a[i][i]*self.equation.residual('all', u, u_nl, fields, fields, bnd)
+            for j in range(i):
+                rhs += self.dt_const*self.a[i][j]*inner(self.k[j], test)*dx
+            self.F.append(mass - rhs)
+
+        # construct variational problems to evaluate tendencies
+        self.k_form = []
+        for i in range(self.n_stages - 1):
+            kf = inner(self.k[i], test)*dx - self.equation.residual('all', u, u_nl, fields, fields, bnd)
+            self.k_form.append(kf)
+
+        self.update_solver()
+
+    def update_solver(self):
+        """Create solver objects"""
+        # Ensure LU assembles monolithic matrices
+        if self.solver_parameters.get('pc_type') == 'lu':
+            self.solver_parameters['mat_type'] = 'aij'
+        self.solver = []
+        for i in range(self.n_stages):
+            p = NonlinearVariationalProblem(self.F[i], self.solution)
+            sname = '{:}_stage{:}_'.format(self.name, i)
+            s = NonlinearVariationalSolver(
+                p, solver_parameters=self.solver_parameters,
+                options_prefix=sname)
+            self.solver.append(s)
+        self.k_solver = []
+        k_solver_parameters = {
+            'snes_type': 'ksponly',
+        }
+        for i in range(self.n_stages - 1):
+            p = NonlinearVariationalProblem(self.k_form[i], self.k[i])
+            sname = '{:}_k_stage{:}_'.format(self.name, i)
+            s = NonlinearVariationalSolver(
+                p, solver_parameters=k_solver_parameters,
+                options_prefix=sname)
+            self.k_solver.append(s)
+
+    def initialize(self, init_cond):
+        """Assigns initial conditions to all required fields."""
+        self.solution_old.assign(init_cond)
+        self._initialized = True
+
+    def get_final_solution(self, additive=False):
+        """
+        Evaluates the final solution
+        """
+        pass
+
+    def solve_stage(self, i_stage, t, update_forcings=None):
+        """
+        Solves a single stage of step from t to t+dt.
+        All functions that the equation depends on must be at right state
+        corresponding to each sub-step.
+        """
+        if i_stage == 0:
+            # NOTE solution may have changed in coupled system
+            self.solution_old.assign(self.solution)
+        if not self._initialized:
+            error('Time integrator {:} is not initialized'.format(self.name))
+        if update_forcings is not None:
+            update_forcings(t + self.c[i_stage]*self.dt)
+        self.solver[i_stage].solve()
+        if i_stage < self.n_stages - 1:
+            self.k_solver[i_stage].solve()
```

Comment:
Storing the tendency here involves another solve which is not ideal. I tried assigning `k = (u^(i) - u_old ...)/dt/a[i][i]` but that doesn't seem to work.

Reply:
Mathematically doesn't work, or implementation craps out?

Reply:
The solver does not converge. I checked the norm of `k` obtained from the solver (as implemented here) and from the assignment and the fields are indeed different. I guess I'm missing something.

Reply:
Well you need the modified mass terms in u^(i) and u_old when doing `k = (u^(i) - u_old ...)/dt/a[i][i]`, i.e. solve (but this is more of a projection):
```
<test, k> = mass_term(u^i) - mass_term(u^0) - sum_{j<i} <test, dt*k_j
```

Reply:
Ah yes, that must be it. The `\tilde h` function does not fit in the function space so it must be projected. Still the projection you proposed is probably cheaper than re-evaluating `equation.residual`.

Reply:
I've added the projection proposed by Stephan in the latest version.